### PR TITLE
Reasoner planner model bypassed reasoning

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## What is the goal of this PR?
Model the fact that the reasoner will bypass reasoning when it is safe to do so.

## What are the changes implemented in this PR?
* Add a method in ReasonerPlanner.triggeredCalls to check if reasoning can be bypassed.
* Make the reasoner planned & concludable controller call this method.
  - Requires the reasoner planner to know if `explain` is enabled - Add constructor parameter & propagate.

### CODE REVIEW

**Comment by benjaminasmall** [comment]
Do you want to preserve any tests where `explain` is `true`?

File: reasoner/planner/ReasonerPlanner.java [hunk 1]
```diff
@@ -128,8 +130,19 @@ private static boolean notGeneratedByResolvable(Resolvable<?> resolvable, Variab
...
 
+    public static boolean canBypassReasoning(Concludable concludable, Set<com.vaticle.typedb.core.traversal.common.Identifier.Variable.Retrievable> boundVariables, boolean isExplainEnabled) {
+        return !isExplainEnabled && !concludable.isHas() &&
+                concludable.generating().isPresent() && boundVariables.contains(concludable.generating().get().id());
+    }
+
     public Set<CallMode> triggeredCalls(Concludable concludable, Set<Variable> mode, @Nullable Set<ResolvableConjunction> dependencyFilter) {
         Set<CallMode> calls = new HashSet<>();
+
+        // Don't trigger reasoning if we can just look it up
```
Comment (flyingsilverfin) [inline reasoner/planner/ReasonerPlanner.java:141]: this comment is inaccurate - the planner doesn't 'trigger' reasoning, it models if it 'will' be triggered in the during execution.

Comment (krishnangovindraj) [reply reasoner/planner/ReasonerPlanner.java]: removed

File: reasoner/planner/ReasonerPlanner.java [hunk 2]
```diff
@@ -128,8 +130,19 @@ private static boolean notGeneratedByResolvable(Resolvable<?> resolvable, Variab
                 .allMatch(constraint -> constraint.isThing() && constraint.asThing().isValue());
     }
 
+    public static boolean canBypassReasoning(Concludable concludable, Set<com.vaticle.typedb.core.traversal.common.Identifier.Variable.Retrievable> boundVariables, boolean isExplainEnabled) {
+        return !isExplainEnabled && !concludable.isHas() &&
+                concludable.generating().isPresent() && boundVariables.contains(concludable.generating().get().id());
+    }
```
[the remark this answers is not in the document]
Comment (krishnangovindraj) [reply reasoner/planner/ReasonerPlanner.java]: Moved it to ConcludableController, though I may not fully understand the second half of the comment.

**Review by flyingsilverfin — APPROVED** [review]
Looks good! Reads much nicer this way around :)

**Pull request merged**